src/tap.c

```diff
@@ -0,0 +1,247 @@
+ *
+ * This file is part of Butterfly.
+ *
+ * Butterfly is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as published
+ * by the Free Software Foundation.
+ *
+ * Butterfly is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <unistd.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <net/if.h>
+#include <linux/if_tun.h>
+#include <arpa/inet.h>
+#include <sys/ioctl.h>
+#include <sys/select.h>
+#include <rte_config.h>
+#include <rte_ethdev.h>
+#include <rte_cycles.h>
+
+#include <packetgraph/packetgraph.h>
+#include "brick-int.h"
+#include "packets.h"
+#include "utils/mempool.h"
+#include "utils/bitmask.h"
+
+struct pg_tap_config {
+	char ifname[IFNAMSIZ];
+};
+
+struct pg_tap_state {
+	struct pg_brick brick;
+	int tap_fd;
+	struct ifreq ifr;
+	struct rte_mbuf *pkts[PG_MAX_PKTS_BURST];
+	/* side of the kernel interface */
+	enum pg_side output;
+};
+
+const char *pg_tap_ifname(struct pg_brick *brick)
+{
+	struct pg_tap_state *state =
+		pg_brick_get_state(brick, struct pg_tap_state);
+
+	return state->ifr.ifr_name;
+}
+
+static struct pg_brick_config *tap_config_new(const char *name,
+					      const char *ifname)
+{
+	struct pg_brick_config *config = g_new0(struct pg_brick_config, 1);
+	struct pg_tap_config *tap_config = g_new0(struct pg_tap_config, 1);
+
+	if (ifname)
+		strncpy(tap_config->ifname, ifname, IFNAMSIZ);
+	config->brick_config = (void *) tap_config;
+	return pg_brick_config_init(config, name, 1, 1, PG_MONOPOLE);
+}
+
+static int tap_burst(struct pg_brick *brick, enum pg_side from,
+		     uint16_t edge_index, struct rte_mbuf **pkts,
+		     uint64_t pkts_mask, struct pg_error **errp)
+{
+	struct pg_tap_state *state = pg_brick_get_state(brick,
+							struct pg_tap_state);
+	uint64_t it_mask;
+	uint64_t bit;
+	uint16_t i;
+	int fd = state->tap_fd;
+
+	it_mask = pkts_mask;
+	for (; it_mask;) {
+		pg_low_bit_iterate_full(it_mask, bit, i);
+		struct rte_mbuf *packet = pkts[i];
+		char *data = rte_pktmbuf_mtod(packet, char *);
+		int len = rte_pktmbuf_pkt_len(packet);
+
+		/* write data. */
+		if (write(fd, data, len) < 0) {
+			*errp = pg_error_new("%s", strerror(errno));
+			return -1;
+		}
+	}
+	return 0;
+}
+
+static int tap_poll(struct pg_brick *brick, uint16_t *pkts_cnt,
+		    struct pg_error **errp)
+{
+	uint16_t nb_pkts;
+	uint64_t pkts_mask;
+	int ret;
+	struct pg_tap_state *state =
+		pg_brick_get_state(brick, struct pg_tap_state);
+	int fd = state->tap_fd;
+	int fd1 = fd + 1;
+	struct pg_brick_side *s = &brick->side;
+	struct rte_mbuf **pkts = state->pkts;
+	static struct timeval timeout;
```

Comment:
I would add a comment to explain that by having a static timeout struct, we're implicitly set the struct to 0 therefore we're aren't waiting in select.